Use govuk components

steventux opened the pull request:
https://trello.com/c/YUaMfA4y/524-add-breadcrumbs-and-related-links-components-to-multipage-frontend-travel-advice-page

Adds breadcrumbs and related links components to multipage frontend...

![screenshot from 2016-02-01 12 18 42](https://cloud.githubusercontent.com/assets/93511/12717170/f425d55a-c8dd-11e5-9f3a-c960d3c16dfa.png)

Depends on https://github.com/alphagov/govuk-content-schemas/pull/227 for test data.

Commit:
Add guide page styles from static

Commit:
Use components instead of static template

Commit:
Add helper method to sort breadcrumbs by parent links.

Commit:
Assign links to Multipage model.

Commit:
Add breadcrumbs component to layout...

Adds a breadcrumb component to the base layout. The data for the breadcrumbs comes
from the content item via the application helper. If no links data is present no breadcrumbs are
rendered.

Adds some support for capybara govuk component matchers.

Commit:
Prepend Home breadcrumb to breadcrumb data.

Commit:
Add a way to format related links data for rendering...

Related links data need ordering and assigning to parent links so that
they can be rendered by the govuk related links component.
Currently the parent links are ordered by the (reverse) breadcrumb precedence,
the top related link parent will be the breadcrumb 'nearest' to the
content item.

fofr commented on app/views/layouts/application.html.erb:
```diff
@@ -13,10 +13,15 @@
   <body>
-    <div id="global-breadcrumb" class="header-context"></div>
     <div id="wrapper" role="main" class="<%= @presenter.content_class_dasherized %> guide">
+      <%= render partial: 'govuk_component/breadcrumbs', locals: { breadcrumbs: breadcrumbs } %>
```
I think the breadcrumbs and related component should be in the `show` template rather than the application one.

fofr commented on app/views/layouts/application.html.erb:
```diff
@@ -13,10 +13,15 @@
 
   <body>
-    <div id="global-breadcrumb" class="header-context"></div>
     <div id="wrapper" role="main" class="<%= @presenter.content_class_dasherized %> guide">
```
A small change here, can we use `<main id="wrapper"...` rather than `<div>`

steventux commented:
thanks for the review @fofr :+1: updated.

fofr commented:
@steventux It looks like you've ported some styles that aren't needed – either because they're already in static or they're provided by components. It's hard to identify exactly what needs removing just from the PR.

There are some other markup changes that would benefit the app, but shouldn't block this PR.

@dsingleton and I were discussing what approach to take. We have a story in the backlog for looking at multipage-frontend: https://trello.com/c/Lv5JExC0/376-multipage-frontend-cleanup. As this isn't a live app yet, we're happy for it to be merged _as is_. When you're happy with how it looks and functions, then we can refine it before it goes live.

bradwright commented:
@fofr we're looking to deploy this as soon as possible. So if you believe there's additional work it needs to be done soon (ideally today).

steventux commented:
@fofr, I've had to port styles in `helpers/_text.scss` from **alphagov/static** and I've taken the entire`views/_travel_advice.scss` from **alphagov/frontend**. Is there a way to import the static helpers to avoid duplication?

tuzz commented on app/helpers/application_helper.rb:
```diff
@@ -2,4 +2,84 @@ module ApplicationHelper
+  def related_links_data
+    @presenter.links[:related] if @presenter.links && @presenter.links.any?
+  end
 end
```
I think this would be made simpler if the presenter encapsulated the breadcrumbs logic.

Right now, the ApplicationHelper needs to know quite a lot about the `links` data structure and business logic of how to order each breadcrumb. I wonder if this would be made clearer if it were handled in the presenter. I'd imagine it'd be easier to test, too. Would be happy to work on this together.

That being said, this works and you could also just ship it with a view to refactoring later.

tuzz commented:
I can't comment too much on the html/css, but the Ruby code looks fine and could ship.

I think there's an improvement that **could** be made around making Breadcrumbs a bit more OO, but this can be shipped to production as-is in my opinion. :+1:

fofr commented:
To clarify after conversation in meatspace – the frontend refactoring won't block this from being merged and deployed.

As it's going live imminently we'll look to pick up the refactoring sooner (but this might be after it's live). When we do we'll work on the assumption that there's no ongoing feature work but there could be bug fixes.

fofr commented:
@steventux Have you tested this looks good cross browser?

Commit:
Render breadcrumbs and related links

Commit:
Add missing print icons

steventux commented:
@tuzz thanks for the review :+1: - I've split the helper method into `BreadcrumbsPresenter` and `RelatedLinksPresenter` and call these from `MultipagePresenter` now, hopefully simpler to follow.

tuzz commented:
Looks good. :+1: from me.

dsingleton commented:
Small point, and not new to this PR but highlighted by some of the spec changes here. 

[Integration tests on government-frontend](https://github.com/alphagov/government-frontend/pull/83) use the [content examples](https://github.com/alphagov/govuk-content-schema-test-helpers), rather than [lots of stubbing](https://github.com/alphagov/multipage-frontend/blob/use_govuk_components/spec/features/show_travel_advice_country_spec.rb#L6-L88), which we foundvery useful. Testing against content examples ensure the frontend is being exercised against real/schema-validated  examples. It worries me that if the content-schema/examples changed significantly, there wouldn't be an associated failing test on this repo.

tuzz commented:
I set the same up in frontend recently. Might be useful as reference:
https://github.com/alphagov/frontend/blob/master/test/integration/travel_advice_test.rb#L16-L17

Commit:
Adjust bottom margin for content-block titles

Commit:
Add govuk_content_schema_test_helpers.

Commit:
Use content schema test helpers to provide content...

Replaces stubbing of content item json with frontend examples from
govuk-content-schemas.

steventux commented:
thanks @tuzz and @dsingleton for test helpers comments, I've added these, this PR now depends on https://github.com/alphagov/govuk-content-schemas/pull/227 to provide data from schemas which should fix failing tests.
